### PR TITLE
ATT-6: Thumbnails should not be generated for (already) small images

### PR DESCRIPTION
**Pull request by oslynn**
Checks original image dimensions. If the with and hight are within
   the THUMBNAIL range, we just append "___nothumb__" to the original file
   name and NOT generate addition thumbnail file.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java [hunk 1]
```diff
@@ -64,23 +68,50 @@ protected boolean deleteComplexData(Obs obs, DocumentComplexData docComplexData)
...
+		               + "OBS_ID='" + obs.getObsId() + "', "
+		               + "FILE='" + docComplexData.getTitle() + "'.", e);
+      }    
+  
+      // Check for small image file
+      if ((imageHeight <= THUMBNAIL_HEIGHT) && (imageWidth <= THUMBNAIL_WIDTH)) { 	  
+    	  // Rename the file by append "_smallfile" to the file.
+    	  // Therefore, we will know this is a small file and no need for thumnail.
+    	  String newSavedFileName = buildSmallFileFileName(savedFile.getAbsolutePath());
+    	  newSavedFile = new File(newSavedFileName);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: `newSavedFile` should be instantiated here.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java [hunk 2]
```diff
@@ -64,23 +68,50 @@ protected boolean deleteComplexData(Obs obs, DocumentComplexData docComplexData)
    @Override
    protected ValueComplex saveComplexData(Obs obs, DocumentComplexData docComplexData) {
 
+	   File newSavedFile = null;
+	   int imageHeight = 0;
+	   int imageWidth = 0;
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: `imageHeight` and `imageWidth` should default to `Integer.MAX_VALUE` (and not `0`), the default behaviour must be to generate a thumbnail.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java [hunk 3]
```diff
@@ -64,23 +68,50 @@ protected boolean deleteComplexData(Obs obs, DocumentComplexData docComplexData)
...
 
       File savedFile = AbstractHandler.getComplexDataFile(obs);
       String savedFileName = savedFile.getName();
       
-      // Saving the thumbnail
-      File dir = savedFile.getParentFile();
-      String thumbnailFileName = buildThumbnailFileName(savedFileName);
+      // Get image dimensions
       try {
-         Thumbnails.of(savedFile.getAbsolutePath()).size(THUMBNAIL_HEIGHT, THUMBNAIL_WIDTH).toFile( new File(dir, thumbnailFileName) );
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: Could you wrap this into a new function
```java
public static void saveThumbnailOrRenameFile(File file, int thumbnailHeight, int thumbnailWidth) {
  ...
}
```
and provide your entire new logic inside this separate new function?

**IMPORTANT**: Provide unit test(s) for this new behaviour in `ImageAttachmentHandlerIT`.

Please ensure that the JavaDoc of this public method details precisely what it does. As in "given an image file saved to disk, either save a thumbnail alongside the original or mark the file as being its own thumbnail with a special suffix."

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java [hunk 4]
```diff
@@ -61,6 +61,10 @@ protected void setParent(ComplexObsHandler complexObsHandler) {
   final protected ComplexObsHandler getParent() {
     return parent;
   }
+  
+  protected static String buildSmallFileFileName(String fileName) {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java]: Refactor to `buildNoThumbnailFileName(..)`. Explain briefly a header comment (`/* ... */`) what this method does.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java [hunk 5]
```diff
@@ -30,7 +32,9 @@ protected void setParentComplexObsHandler() {
    protected ComplexData readComplexData(Obs obs, ValueComplex valueComplex, String view) {
       
       String fileName = valueComplex.getFileName();
-      if (view.equals(AttachmentsConstants.ATT_VIEW_THUMBNAIL)) {
+      
+      if (view.equals(AttachmentsConstants.ATT_VIEW_THUMBNAIL) && 
+    		  !fileName.contains("_smallfile")) {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: Given the comment above, this check will go away.

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java [hunk 6]
```diff
@@ -61,6 +61,10 @@ protected void setParent(ComplexObsHandler complexObsHandler) {
...
+      return FilenameUtils.removeExtension(fileName) + "_smallfile" + "." + FilenameUtils.getExtension(fileName);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java]: `_smallfile` should be maintained as a `final static String`, either `public` or `protected`.
I would favor something like:
```
public final static String NO_THUMBNAIL_SUFFIX = "___nothumb__";
```
The multiple underscores will guarantee a "certain" uniqueness. Not perfect, but good enough.

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java [hunk 7]
```diff
@@ -61,6 +61,10 @@ protected void setParent(ComplexObsHandler complexObsHandler) {
...
+  }
 
   protected static String buildThumbnailFileName(String fileName) {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java]: This method should check whether the special "no thumbnail" suffix is part of the name, and if it is part of the name then `fileName` should be returned. That would fulfil the concept that "the image is its own thumbnail" (if it is marked as such.)

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java (hunk 5 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: Note that `deleteComplexData(..)` will have to be amended as well...

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java (hunk 1 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/attachments/obs/ImageDocumentHandler.java]: Perhaps `renamedFile` is a more explicit naming?

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java (hunk 7 above)
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/attachments/obs/AbstractDocumentHandler.java]: **ATTENTION**: It is important that the check searches for a _suffix_ and not for the file name to just contain the special string. This is because we cannot ever be sure that a file containing the special string might ever be saved in the first place (in fact it is totally possible). So basically: `StringUtils.endsWith(str, suffix)` ;-)

**Review by mks-d — CHANGES_REQUESTED** [review]
Hi @oslynn . Thanks so much for looking into it.
I have left quite a few comments. The hardest piece will be to bring new unit tests in the new `ImageAttachmentHandlerIT`.

I like a lot your idea of renaming the file to mark it as a file with no thumbnail. The method has its edge cases (see my comments) but it's quite nifty!

You will have to do a 
```
git pull --rebase origin master
```

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java [hunk 8]
```diff
@@ -51,39 +53,55 @@ protected boolean deleteComplexData(Obs obs, AttachmentComplexData complexData)
...
+	    boolean isThumbNailPurged = false;
 		Obs tmpObs = new Obs();
-		tmpObs.setValueComplex(thumbnailFileName);
-		boolean isThumbNailPurged = getParent().purgeComplexData(tmpObs);
+
+	    if (!StringUtils.endsWith(FilenameUtils.removeExtension(fileName), NO_THUMBNAIL_SUFFIX)) {
+	    	thumbnailFileName = buildThumbnailFileName(fileName);
+			tmpObs.setValueComplex(thumbnailFileName);
+			isThumbNailPurged = getParent().purgeComplexData(tmpObs);
+	    }
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java]: Add and else statement that says that the thumbnail is purged (no thumbnail means it's already purged):
```
else {
  isThumbNailPurged = true;
}
```

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java [hunk 9]
```diff
@@ -51,39 +53,55 @@ protected boolean deleteComplexData(Obs obs, AttachmentComplexData complexData)
...
+		
 		tmpObs.setValueComplex(fileName);
 		boolean isImagePurged = getParent().purgeComplexData(tmpObs);
 		
-		return isThumbNailPurged && isImagePurged;
+		return isThumbNailPurged || isImagePurged;
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java]: ... and revert to the original AND statement.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java [hunk 10]
```diff
@@ -51,39 +53,55 @@ protected boolean deleteComplexData(Obs obs, AttachmentComplexData complexData)
...
-		String thumbnailFileName = buildThumbnailFileName(fileName);
-		
+	    String thumbnailFileName = null;
+	    boolean isThumbNailPurged = false;
 		Obs tmpObs = new Obs();
-		tmpObs.setValueComplex(thumbnailFileName);
-		boolean isThumbNailPurged = getParent().purgeComplexData(tmpObs);
+
+	    if (!StringUtils.endsWith(FilenameUtils.removeExtension(fileName), NO_THUMBNAIL_SUFFIX)) {
+	    	thumbnailFileName = buildThumbnailFileName(fileName);
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java]: `thumbnailFileName`should be instantiated here and not on line 56, it is not needed outside the `if` scope.

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractAttachmentHandler.java [hunk 11]
```diff
@@ -61,11 +70,60 @@ protected void setParent(ComplexObsHandler complexObsHandler) {
...
+	* @param imageHight image hight
+	* @param imageWidth image width
+	* @return savedFileName new renamed file name or orignal file name
+	*/
+	public static String saveThumbnailOrRename(File savedFile, int imageHight, int imageWidth){
+
+       String savedFileName = savedFile.getName();
+
+       // Check for small image file
+       if ((imageHight <= THUMBNAIL_MAX_HEIGHT) && (imageWidth <= THUMBNAIL_MAX_WIDTH)){
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/AbstractAttachmentHandler.java]: `imageHight` --> `imageHeight`

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 12]
```diff
@@ -22,7 +24,7 @@
 	protected TestHelper testHelper;
 
 	@Test
-	public void saveComplexData_shouldSaveThumbnailToDisk() throws IOException {
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: @oslynn please leave the existing test untouched. However you will need to bring new test cases that are specific to the case of small images.

**Test case 1**: `saveComplexData_shouldNotSaveThumbnailWhenSmallImage`
This test case will replay the saving of an image, but a _small_ image whose original dimensions are under 200 x 200. And it will have to verify that when saving such image, no new thumbnail is generated and the original image file is renamed. 

**Test case 2**: `readComplexData_shouldAlwaysFetchOriginalImageWhenSmallImage`
We want to check that we fetch the original image, whether we ask for the thumbnail or not, when we are in the case of a small image.

You will need `TestHelper` to cover that case with a new, small resource. I can do that for you in an upcoming commit of mine.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java [hunk 13]
```diff
@@ -51,39 +53,55 @@ protected boolean deleteComplexData(Obs obs, AttachmentComplexData complexData)
...
 		obs = getParent().saveObs(obs);
-		
+
 		File savedFile = AbstractHandler.getComplexDataFile(obs);
-		String savedFileName = savedFile.getName();
-		
-		// Saving the thumbnail
-		File dir = savedFile.getParentFile();
-		String thumbnailFileName = buildThumbnailFileName(savedFileName);
+		// String savedFileName = savedFile.getName();
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java]: I think this was better, please keep this instantiation and move it to just above line 96.

File: api/src/main/java/org/openmrs/module/attachments/obs/AbstractAttachmentHandler.java [hunk 14]
```diff
@@ -61,11 +70,60 @@ protected void setParent(ComplexObsHandler complexObsHandler) {
...
+    	  // Rename the file by append "_smallfile" to the file.
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/AbstractAttachmentHandler.java]: Became `NO_THUMBNAIL_SUFFIX`.

**Review by mks-d — CHANGES_REQUESTED** [review]
Thanks @oslynn , almost there. You will have to bear with me for extending `TestHelper` (see my comments about further unit tests). Otherwise I have some other smaller comments that you can already address through a next commit.

On a milder note there is an indentation issue. This could come from using different configurations on different IDEs (very annoying). Could you try to look into what is causing this? I will on my side try to apply an automatic code formatter for the Maven build so that things get straightened automatically independently of the developer who's been working on the code.

File: api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java [hunk 15]
```diff
@@ -14,94 +14,92 @@
...
+        obs = getParent().saveObs(obs);
+
+        File savedFile = AbstractHandler.getComplexDataFile(obs);
+
+        // Get image dimensions
+        try {
+            BufferedImage image = ImageIO.read(savedFile);
+            imageHeight = image.getHeight();
+            imageWidth = image.getWidth();
+        } catch (IOException e) {
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/attachments/obs/ImageAttachmentHandler.java:87]: I have been thinking here. If the dimensions could not be determined, then you should not throw an exception but let the code go on with the maximum dimensions initially set. Then after that if the thumbnail generation fails, _then_ so be it. So basically just:
```java

try {
  BufferedImage image = ImageIO.read(savedFile);
  imageHeight = image.getHeight();
  imageWidth = image.getWidth();
} catch (IOException e) {
  log.warn("The dimensions of image 'savedFileName' could not be determined, continuing with generating thumbnail.", e)  
}
```

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 16]
```diff
@@ -20,63 +20,133 @@
...
-		// Verif
-		Assert.assertArrayEquals(expectedBytes, BaseComplexData.getByteArray(obs.getComplexData()));
-	}
-}
\ No newline at end of file
+    @Autowired
+    protected TestHelper testHelper;
+
+    @Test
+    public void saveComplexData_shouldSaveThumbnailToDiskAndRename() throws IOException {
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: You should revert its original name, but this will happen anyway when rebasing.

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 17]
```diff
@@ -20,63 +20,133 @@
...
+        Obs obs = testHelper.saveImageAttachment("smallSize");
+        ValueComplex vc = new ValueComplex(obs.getValueComplex());
+        
+        // Get original file name
+        String originalFileName = StringUtils.strip(
+                FilenameUtils.removeExtension(vc.getFileName()), ImageAttachmentHandler.NO_THUMBNAIL_SUFFIX) + 
+                "." + FilenameUtils.getExtension(vc.getFileName());        
+
+        // Verif the buildNotThumbnailFileFileName()
+        File nothumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildNoThumbnailFileFileName(originalFileName));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: Name this variable `noThumbnailFile`.

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 18]
```diff
@@ -20,63 +20,133 @@
...
+        
+        // Verif to thumbnail file was NOT created
+        File thumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildThumbnailFileName(originalFileName));
+        Assert.assertFalse(thumbnail.exists());
+
+        //Assert.assertThat(thumbnail.length(), lessThan(file.length()));
+        BufferedImage img = ImageIO.read(nothumbnail);
+        Assert.assertEquals(ImageAttachmentHandler.THUMBNAIL_MAX_HEIGHT, Math.max(img.getHeight(), img.getWidth()));
+
+        // Now check rename the file by appending NO_THUMBNAIL_SUFFIX to the file
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: Simply say "Check that the file name contains the no thumbnail suffix."

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 19]
```diff
@@ -20,63 +20,133 @@
...
+        Assert.assertTrue(StringUtils.endsWith(FilenameUtils.removeExtension(nothumbnail.getName()), ImageAttachmentHandler.NO_THUMBNAIL_SUFFIX));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: +1 Exactly what was needed.

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 20]
```diff
@@ -20,63 +20,133 @@
...
+
+        // Verif the buildNotThumbnailFileFileName()
+        File nothumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildNoThumbnailFileFileName(originalFileName));
+        Assert.assertTrue(nothumbnail.exists());
+        // Not create newe file name since it contained NO_THUMBNAIL_SUFFIX
+        nothumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildNoThumbnailFileFileName(vc.getFileName()));
+        Assert.assertTrue(nothumbnail.exists());        
+        
+        // Verif to thumbnail file was NOT created
+        File thumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildThumbnailFileName(originalFileName));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: +1 Exactly what was needed.

File: api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java [hunk 21]
```diff
@@ -20,63 +20,133 @@
...
+                "." + FilenameUtils.getExtension(vc.getFileName());        
+
+        // Verif the buildNotThumbnailFileFileName()
+        File nothumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildNoThumbnailFileFileName(originalFileName));
+        Assert.assertTrue(nothumbnail.exists());
+        // Not create newe file name since it contained NO_THUMBNAIL_SUFFIX
+        nothumbnail = new File(testHelper.getComplexObsDir() + "/" + ImageAttachmentHandler.buildNoThumbnailFileFileName(vc.getFileName()));
+        Assert.assertTrue(nothumbnail.exists());        
+        
+        // Verif to thumbnail file was NOT created
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java]: `to` -> `the`

**Review by mks-d — CHANGES_REQUESTED** [review]
Ok first of all this is good! Your unit tests are spot on.

However you will need to rebase first on master for an important reason: I had actually improved the existing tests by extending `TestHelper` API (see [here](https://github.com/openmrs/openmrs-module-attachments/blob/b8c24b45eca9de7bb62e4213f3e9504f86588f04/api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java#L63-L68)). I am not using the `ValueComplex` instance anymore to extract the original file name (that was a weakness), I refer to the original file exposed by the API. See [here](https://github.com/openmrs/openmrs-module-attachments/blob/b8c24b45eca9de7bb62e4213f3e9504f86588f04/api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerIT.java#L32) for example.

I'm sorry there's been an overlap because I needed to extend the test framework ahead of your work. I can see that you came up with changes that are very similar to what I have done in the last commit which is great as well. However the bottom line is always: `git pull --rebase origin master`. The result here is that you have conflicting files in the end, so it cannot be merged in anyway.

Last thing: could you have a look at the formatting changes that you introduce? I'm trying to get a clue as to what is happening, perhaps could you inspect with a standard text editor (say Sublime) a file that you have modified in your IDE. More [here](https://talk.openmrs.org/t/code-formatting-upon-building-a-module/14300) about this.

**Comment by oslynn** [comment]
Done, ATT-6 merged to the master.
